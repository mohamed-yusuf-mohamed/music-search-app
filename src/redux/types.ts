export interface ResponseData {
  artistName: string,
  kind: string,
  trackName: string, 
  artworkUrl60: string   
}

// TODO: types
export interface SearchBarState {
  loading: boolean,
  error: boolean,
  // status: 'loading' | 'loaded' | 'error',
  parsedInput: string,
  input: string,
  data: ResponseData[],
  fetchCount: number
}

const initialState: SearchBarState = {
  // status: ""
  loading: false,
  error: false,
  parsedInput: "",
  input: "",
  data: [],
  fetchCount: 1
};