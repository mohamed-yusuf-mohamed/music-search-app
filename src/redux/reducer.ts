// import { combineReducers } from 'redux'
import {ResponseData} from "./types"

// COUNTER REDUCER
// const counterReducer = (state = 0, { type }) => {
//   switch (type) {
//     case types.INCREMENT:
//       return state + 1
//     case types.DECREMENT:
//       return state - 1
//     case types.RESET:
//       return 0
//     default:
//       return state
//   }
// }

// INITIAL TIMER STATE
const initialState = {
    loading: false,
    parsedInput: "",
    input: "",
    data: [] as ResponseData[],
    fetchCount: 0,
    error: false
}

// TIMER REDUCER
// TODO: types
// TODO: refactor
const reducer = (state = initialState, action: any) => {
  // ...state.data
  // ...action.payload.data
  // action.payload.data
  
  switch (action.type) {

    
    case "FETCH_STARTED":
      return {
        ...state,
        loading: true
        // fetchCount: state.fetchCount + 1
      }
    case "TEST":
      return {
        ...state,
        fetchCount: state.fetchCount + 1,
        data: action.payload.data
        // fetchCount: state.fetchCount + 1
      }
    case "FETCH_SUCCESS":
      return {
        ...state,
        data: action.payload.data,
        // fetched: action.payload.data,
        loading: false,
        fetchCount: state.fetchCount + 1
      }
    case "FETCH_ERROR":
      return {
        ...state,
        error: true,
        loading: false,
        fetchCount: state.fetchCount - 1
      }
    case "CHANGE_INPUT":
      return {
        ...state,
        input: action.payload,
        parsedInput: action.payload.split(' ').join('+')
      }
    default:
      return state
  }
}

// COMBINED REDUCERS
// const reducers = {
//   counter: counterReducer,
//   timer: timerReducer,
// }

export default reducer

// export default combineReducers(reducers)
