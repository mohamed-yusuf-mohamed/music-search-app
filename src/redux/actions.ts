import Api from "../api"
import {Dispatch, State, GetState, Thunk} from "./store"
import { Action } from '@reduxjs/toolkit';
import { AnyAction } from '@reduxjs/toolkit';

// TODO: change commit author and name on all commits

const api = new Api()

export const fetchData = (): Thunk => async (dispatch: Dispatch, getState: GetState) => {
    const state = getState()
    const {parsedInput, page, data} = state
    const offset = page * 10
    // TODO: validate the offset is working as it should
    const url = `https://itunes.apple.com/search?term=${parsedInput}&offset=${offset}&limit=10`

    try {
      const res = await api.fetch(url)
      return dispatch({
        type: "FETCH",
        payload: {
          data: {
            ...data,
            ...res
          }
        }
      })
    } catch(err) {
      // TODO: error details to store
      return dispatch({
        type: "ERROR",
        payload: {
          error: true,
          loading: false
        }
      })
    }
}

export const handleInput = (e: React.ChangeEvent<HTMLInputElement>): AnyAction => ({ 
  type: "INPUT", 
  payload: e.target.value 
})