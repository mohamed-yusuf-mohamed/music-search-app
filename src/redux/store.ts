import { configureStore, ThunkAction, Action, applyMiddleware } from '@reduxjs/toolkit';
import { combineReducers } from 'redux'
// import counterReducer from '../../DELETEME/counter/counterSlice';
import reducer from "./reducer"
import thunk from "redux-thunk"
import { createStore } from 'redux';

// const reducer = combineReducers({counterReducer, appReducer})
// import rootReducer from './reducers/index';

// Note: this API requires redux@>=3.1.0
// export const store = createStore(reducer, applyMiddleware(thunk));

export const store = createStore(reducer, applyMiddleware(thunk))

// export const store = configureStore({
//   reducer: {
//     counter: counterReducer,
//     app: appReducer
//   }
// });

// export const 

export type AppDispatch = typeof store.dispatch;
export type RootState = ReturnType<typeof store.getState>;
export type AppThunk<ReturnType = void> = ThunkAction<
  ReturnType,
  RootState,
  unknown,
  Action<string>
>;
