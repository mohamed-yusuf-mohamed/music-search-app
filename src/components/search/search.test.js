import React from 'react';
import { render, fireEvent } from '@testing-library/react';
import SearchInput from "./index"
import {fetchData, handleInput} from "../../redux/actions";
import { Provider } from 'react-redux';
import { store } from '../../redux/store';
import userEvent from '@testing-library/user-event'

jest.mock("../../redux/actions", () => ({
  ...jest.requireActual("../../redux/actions"),
  fetchData: jest.fn(() => ({type: ""}))
}))

describe("search input component", () => {
  const {getByTestId} = render(
    <Provider store={store}>
        <SearchInput />
    </Provider>
  );
  const input = document.querySelector("[data-testid=search-component] input")
  const component = getByTestId("search-component")

  it('renders', () => {
    expect(component).toMatchSnapshot()
  });

  it("on input change", () => {
    userEvent.type(input, "apples oranges")
    expect(input.value).toBe("apples oranges")
  })

  it.only("on pressing enter", () => {
    userEvent.type(input, "bananas grapes")
    userEvent.keyboard("{Enter}")
    expect(fetchData).toHaveBeenCalled()
  })
})
