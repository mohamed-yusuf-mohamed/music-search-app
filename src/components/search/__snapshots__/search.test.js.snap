// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`search input component renders 1`] = `
<div
  class="MuiFormControl-root MuiTextField-root dbawFk"
  data-testid="search-component"
>
  <div
    class="MuiInputBase-root MuiOutlinedInput-root MuiInputBase-formControl MuiInputBase-adornedStart MuiOutlinedInput-adornedStart"
  >
    <div
      class="MuiInputAdornment-root MuiInputAdornment-positionStart"
    >
      <svg
        aria-hidden="true"
        class="MuiSvgIcon-root"
        focusable="false"
        viewBox="0 0 24 24"
      >
        <path
          d="M15.5 14h-.79l-.28-.27c1.2-1.4 1.82-3.31 1.48-5.34-.47-2.78-2.79-5-5.59-5.34-4.23-.52-7.79 3.04-7.27 7.27.34 2.8 2.56 5.12 5.34 5.59 2.03.34 3.94-.28 5.34-1.48l.27.28v.79l4.25 4.25c.41.41 1.08.41 1.49 0 .41-.41.41-1.08 0-1.49L15.5 14zm-6 0C7.01 14 5 11.99 5 9.5S7.01 5 9.5 5 14 7.01 14 9.5 11.99 14 9.5 14z"
        />
      </svg>
    </div>
    <input
      aria-invalid="false"
      class="MuiInputBase-input MuiOutlinedInput-input MuiInputBase-inputAdornedStart MuiOutlinedInput-inputAdornedStart"
      placeholder="Search for your favourite song, artist or album"
      type="text"
      value=""
    />
    <fieldset
      aria-hidden="true"
      class="PrivateNotchedOutline-root-1 MuiOutlinedInput-notchedOutline"
      style="padding-left: 8px;"
    >
      <legend
        class="PrivateNotchedOutline-legend-2"
        style="width: 0.01px;"
      >
        <span>
          â€‹
        </span>
      </legend>
    </fieldset>
  </div>
</div>
`;
